### PR TITLE
Create statistics of closed list of speakers

### PR DESCRIPTION
**Pull request by GabrielInTheWorld**
AP 10

### CODE REVIEW

File: client/src/app/site/common/components/user-statistics/user-statistics.component.ts [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        for (const structureLevel of this.speakersOfStructureLevel.keys()) {
+            const structureLevelObject = this.speakersOfStructureLevel.get(structureLevel);
+            structureLevelObject.durationString = this.parseDuration(this.getTime(structureLevelObject.durationNumber));
+            this.speakersOfStructureLevel.set(structureLevel, structureLevelObject);
+        }
+        this.speakersOfStructureLevelSubject.next(Array.from(this.speakersOfStructureLevel.values()));
+        this.cd.detectChanges();
+    }
+
+    private addUniqueSpeaker(speaker: ViewSpeaker): boolean {
```
Comment (tsiegleauq) [inline client/src/app/site/common/components/user-statistics/user-statistics.component.ts:161]: the `list-of-speaker-repository` can basically get this number for you.
see `isFirstContribution` as a reference. Basically, get all finishedSpeakers in all lists, map them to their unique user Id's and create an `set` using `[...new Set(myArray)];`

Try to use existing services and avoid putting too much logic into components.

File: client/src/app/site/common/components/user-statistics/user-statistics.component.ts [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        this.statisticIsOpen = !this.statisticIsOpen;
+    }
+
+    /**
+     * This iterates over a list of list-of-speakers. For each speaker it calculates the duration the speaker
+     * has spoken.
+     *
+     * @param listOfSpeakers List of list-of-speakers.
+     */
+    private calcWordRequests(listOfSpeakers: ViewListOfSpeakers[]): void {
```
Comment (tsiegleauq) [inline client/src/app/site/common/components/user-statistics/user-statistics.component.ts:132]: Should also be handled of the list-of-speaker repo. Basically the length of finished-speakers.

File: client/src/app/site/common/components/user-statistics/user-statistics.component.ts (hunk 1 above)
Comment (GabrielInTheWorld) [reply client/src/app/site/common/components/user-statistics/user-statistics.component.ts]: I solved it in a slightly different way. A speaker is pushed immediately to a resulting array, if not already existing.

**Comment by tsiegleauq** [comment]
Structure level should better not show undefined
image

File: client/src/app/core/repositories/agenda/list-of-speakers-repository.service.ts [hunk 3]
```diff
@@ -228,6 +239,96 @@ export class ListOfSpeakersRepositoryService extends BaseHasContentObjectReposit
...
+        }
+        return firstContributions;
+    }
+
+    /**
+     * Maps structure-level to speaker.
+     *
+     * @returns A map, which keys are structure-levels and values a `SpeakingTimeStructureLevelObject`.
+     */
+    public getSpeakingTimeStructureLevelRelation(): Map<string, SpeakingTimeStructureLevelObject> {
```
Comment (FinnStutzenstein) [inline client/src/app/core/repositories/agenda/list-of-speakers-repository.service.ts:265]: This function is just used once and only the values are needed. So you can return just a list of `SpeakingTimeStructureLevelObject`

**Pull request merged**